### PR TITLE
[8.0] add account_invoice_product_supplier_price_update

### PR DESCRIPTION
**Pull request by chafique-delli**
This module allows in the supplier invoice, to automatically update all products whose unit price on the line is different from the supplier price.

*  **account_invoice_supplierinfo_update** : provides main features to update supplierinfo from invoice line information
* **account_invoice_supplierinfo_update_discount** : glue module to update discount field, if 'product_supplierinfo_discount' is installed.
* **account_invoice_supplierinfo_update_variant** : glue module to update variant supplierinfo, if 'product_variant_supplierinfo' is installed.
*  **account_invoice_supplierinfo_update_on_validate** : extra module to force user to update supplierinfo data when user validates the invoice.

### CODE REVIEW

File: account_invoice_product_supplier_price_update/README.rst
Comment (elicoidal) [inline account_invoice_product_supplier_price_update/README.rst]: You can simplify this part with the newest template

File: account_invoice_product_supplier_price_update/__openerp__.py
Comment (elicoidal) [inline account_invoice_product_supplier_price_update/__openerp__.py]: In the supplier invoice, automatically update all products 
whose unit price on the line is different from the supplier price

File: account_invoice_product_supplier_price_update/__openerp__.py
Comment (elicoidal) [inline account_invoice_product_supplier_price_update/__openerp__.py]: probably not necessary as this is the default one

**Comment by chafique-delli** [comment]
@elicoidal , thank you for the code review, I fixed this.

File: account_invoice_product_supplier_price_update/README.rst
Comment (eLBati) [inline account_invoice_product_supplier_price_update/README.rst]: `created` --> `creates`

**Comment by chafique-delli** [comment]
@eLBati , thank you for the code review, I fixed this.

File: account_invoice_product_supplier_price_update/README.rst
Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/README.rst]: s/allow/allows

Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/README.rst]: s/, automatically/, to automatically

File: account_invoice_product_supplier_price_update/__openerp__.py
Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/__openerp__.py]: This module and the following are not real dependencies. Just conveniences, but the module will work the same, except for that features.

File: account_invoice_product_supplier_price_update/__openerp__.py
Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/__openerp__.py]: purchase is not needed for having product.supplierinfo. They are in the product module.

File: account_invoice_product_supplier_price_update/__openerp__.py
Comment (chafique-delli) [inline account_invoice_product_supplier_price_update/__openerp__.py]: @pedrobaeza ,  thank you for the code review.
Indeed, the dependencies on the modules 'purchase' and 'product_variant_supplierinfo' are not necessary.

Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/__openerp__.py]: And you don't need also `product_supplierinfo_tree_price_info`

Comment (chafique-delli) [inline account_invoice_product_supplier_price_update/__openerp__.py]: @pedrobaeza ,in my code, I use the field 'unit_price note' of 'product.supplierinfo' which comes from the module 'product_supplierinfo_tree_price_info'.

Comment (pedrobaeza) [inline account_invoice_product_supplier_price_update/__openerp__.py]: But that module is only for displaying information in an human readable form, not for using for another purposes. I would say to not using anything from this module.

**Comment by chafique-delli** [comment]
Travis test is red because the test of  'account_invoice_merge_purchase' module failed.

File: account_invoice_product_supplier_price_update/account_invoice.py
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/account_invoice.py]: you must depend on this module to get the right supplier https://github.com/OCA/product-variant/pull/32

**Comment by sebastienbeau** [comment]
@chafique-delli can you fis the way to get the right supplier and add a test with
- a specific supplier (a supplier info attach to one variant)
- a generic supplier (a supplier info attach to the template => all variant)

File: account_invoice_product_supplier_price_update/account_invoice.py
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/account_invoice.py]: The code can be more simple here.
- first the current_price_unit do not need the 6, 0 as you always use the [0][2] after so you can remove it.
- then you do not need all of the variable suppinfo...

`
something like that:

```
def _check_product_supplierprice(self):                         
lines = []                                                             
for line in self.invoice_line:                                         
    current_price_unit = []
    supplier = self.partner_id.commercial_partner_id or self.partner_id
suppinfo = False
    current_price_uni = []
    for seller in line.product_id.seller_ids:                          
        if supplier == seller.name: 
            suppinfo = seller
            current_price_unit = [pricelist.price for p in suppinfo.pricelist_ids]
            break                                                      
    if line.price_unit not in current_price_unit:                                               
        lines.append((0, 0, {                                          
            'name': line.name,                                         
            'current_price_unit': suppinfo.unit_price_note,            
            'new_price_unit': line.price_unit,                         
            'suppinfo_id': suppinfo.id,                                
            'to_variant': True,                                        
            'supplier_id': supplier_id.id,                             
            'product_id': line.product_id.id,                          
            'product_tmpl_id': line.product_id.product_tmpl_id.id,     
        }))                                                            
return lines                                                           
```

`

File: account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py]: can you found a shorter class name?

File: account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py]: can you avoid the repetition of self.env['product.supplierinfo'] by creating a variable called supinfo_obj ?

File: account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py]: can you find a shorter class name?

**Review by sebastienbeau — CHANGES_REQUESTED** [review]
Some code simplification should be done

File: account_invoice_product_supplier_price_update/tests/test_supplier_invoice_update_product_supplier_price.py
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/tests/test_supplier_invoice_update_product_supplier_price.py]: We should split the test in 3 as there is 3 test on 3 invoice

**Review by sebastienbeau — COMMENTED** [review]
split test

File: account_invoice_product_supplier_price_update/tests/test_update_supplier_price.py
Comment (sebastienbeau) [inline account_invoice_product_supplier_price_update/tests/test_update_supplier_price.py]: Can you split the test ? one per invoice

**Comment by chafique-delli** [comment]
@gurneyalex, Is you can watch why the runbot test shows me this error:
except_orm: ('Error', u"You try to install module 'account_invoice_product_supplier_price_update' that depends on module 'product_variant_supplierinfo'.\nBut the latter module is not available in your system.")
While the file 'oca_dependencies.txt' contains' product-variant' repository which contains  product_variant_supplierinfo' module.
Thank you.

**Comment by gurneyalex** [comment]
@chafique-delli runbot should be fixed now.

**Comment by chafique-delli** [comment]
@gurneyalex , thank you for the fix on runbot, but is you can watch why the travis test shows an error on installing the  account_invoice_merge_purchase module whereas,  a priori my  account_invoice_product_supplier_price_update module is not installed yet.

**Comment by gurneyalex** [comment]
@chafique-delli this is not your fault. The 8.0 branch build is red on Travis (https://travis-ci.org/OCA/account-invoicing/jobs/166920774)  -> #185

File: account_invoice_product_supplier_price_update/wizard/update_supplier_price.py
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/wizard/update_supplier_price.py]: Why do you launch this workflow ?

File: account_invoice_product_supplier_price_update/wizard/update_supplier_price.py
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/wizard/update_supplier_price.py]: self.ensure_one() is required if you do not manage a for each wizard in self.

File: account_invoice_product_supplier_price_update/wizard/update_supplier_price.py
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/wizard/update_supplier_price.py]: maybe calling default value function could allow to have different user defined default value ? (not sure about that point, just a question).

File: account_invoice_product_supplier_price_update/__openerp__.py
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/__openerp__.py]: Hi. I understand why you depend on this module, but for me, the feature should work if product_variant_supplierinfo is not installed.
What about to split the module in two : 
-  account_invoice_product_supplier_price_update that will depends on account only and work with generic feature
- account_invoice_variant_product_supplier_price_update that will depends on 'product_variant_supplierinfo' and 'account_invoice_product_supplier_price_update', managing variants specifics, overloading function.

If people are agree with that point, I can help you for that refactoring.

File: account_invoice_product_supplier_price_update/__openerp__.py
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/__openerp__.py]: I don't see why you depend on this module. (even if this module is cool ;-) )

File: account_invoice_product_supplier_price_update/__openerp__.py
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/__openerp__.py]: could you git mv for
- demo/product_supplierinfo.xml
- demo/account_invoice.xml

[Ref](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#file-naming)

File: account_invoice_product_supplier_price_update/account_invoice.py
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/account_invoice.py]: something like that should more rapid:

```
for seller in line.product_id.seller_ids.filtered(lambda seller: seller.name== supplier):
```

File: account_invoice_product_supplier_price_update/account_invoice.py
Comment (legalsylvain) [inline account_invoice_product_supplier_price_update/account_invoice.py]: Not a clear idea here, but I think we should think this algorithm as overloadable.

I think to [product_supplierinfo_discount](https://github.com/OCA/purchase-workflow/tree/10.0/product_supplierinfo_discount)

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Hi, @chafique-delli. Some remarks inline.  2 others below. Just a code review for the time being, I'll test it functionaly and more deeply friday.

Otherwise, thanks a lot to share this work !
For shops it is a "must have feature".
- Not sure the image module will help people to understand what the module makes. Default OCA module should be better ? (not a blocking point for me).
- Not sure to have understand all the workflow stuff : invoice validation... (but I just do a quick code review, I'll take a time friday).

**Comment by legalsylvain** [comment]
Hi @chafique-delli,
I took a deeply look on your PR. Some remarks :
* This module should be renamed better in account_invoice_supplierinfo_update because it's not only the price that is updated. (min_qty should be too)
* I still think that it should not depend on variant settings because it adds an unncessary dependency. It should be managed by glue module. The module you're sharing is very generic and is a must have feature, and supplierinfo for variant is specific. The current design will avoid some guys to install this module, or to have to disable manually some undesired features.
* This module doesn't work in all situation. Specially if there are some supplierinfo defined. (min_qty is not managed correctly).

I'm working on a PR on your PR, refactoring your PR. 

Thanks.

**Comment by legalsylvain** [comment]
hi all, 

FYI

If you're interested by this module, a pending refactoring is available here. 
https://github.com/akretion/account-invoicing/pull/3

**Review by legalsylvain — APPROVED** [review]

File: oca_dependencies.txt
Comment (moylop260) [inline oca_dependencies.txt]: Where are required these repositories?

**Comment by legalsylvain** [comment]
Hi. 
I just rebased this PR, after the last merge. 
@chafique-delli, could you squach your commits ? (I can make it if you want).
@sebastienbeau, @all could you review this PR updated ?

Thanks !

**Comment by chafique-delli** [comment]
@legalsylvain , I want you to do it.Thank you.

File: oca_dependencies.txt
Comment (legalsylvain) [reply oca_dependencies.txt]: Thanks for pointing this unused dependencies. I just fixed it.

**Comment by legalsylvain** [comment]
Note : Travis is red because upstream 8.0 branch is red.

**Review by sebastienbeau — APPROVED** [review]

**Comment by sebastienbeau** [comment]
Great work @legalsylvain @chafique-delli  ! Thx

**Comment by legalsylvain** [comment]
This PR has been strongly refactored and is now validated by 2 people. ( @sebastienbeau and me), but as I wrote a lot of the refactored code, I'd prefer to have a full external :+1: 

@eLBati , @pedrobaeza , @elicoidal , @moylop260 : is someone interested to finish to review that PR that you began to review ?

kind regards.

File: account_invoice_supplierinfo_update/models/account_invoice_line.py
Comment (moylop260) [inline account_invoice_supplierinfo_update/models/account_invoice_line.py]: If you have a `if-return` then the `else` is not needed

File: account_invoice_supplierinfo_update/models/account_invoice_line.py
Comment (moylop260) [inline account_invoice_supplierinfo_update/models/account_invoice_line.py]: Unnecessary parents `(`

File: account_invoice_supplierinfo_update/wizard/wizard_update_invoice_supplierinfo_line.py
Comment (moylop260) [inline account_invoice_supplierinfo_update/wizard/wizard_update_invoice_supplierinfo_line.py]: In order to reduce cyclomatic complexity (and quantity of tests to cover the code) you could remove `if-else` sentence using filtered and global write:
```python
self.write({'price_variation': False})
for line in self.filtered('current_price'):
     line.price_variation = 100 * (line.new_price - line.current_price) / line.current_price
```

File: account_invoice_supplierinfo_update_discount/models/account_invoice_line.py
Comment (moylop260) [inline account_invoice_supplierinfo_update_discount/models/account_invoice_line.py]: new api supports BrowseNull.value <- `False`
Use: `res['current_discount'] = partnerinfo.discount`

File: account_invoice_supplierinfo_update_variant/wizard/wizard_update_invoice_supplierinfo_line.py
Comment (moylop260) [inline account_invoice_supplierinfo_update_variant/wizard/wizard_update_invoice_supplierinfo_line.py]: https://github.com/OCA/account-invoicing/blob/d0987a68bfc9c86a59e811c6262327d71cbafcb1/account_invoice_supplierinfo_update/wizard/wizard_update_invoice_supplierinfo_line.py#L62-L69

The original method have the same values but changing just `product_id`:
```python
    @api.multi
    def _prepare_supplierinfo(self):
        self.ensure_one()
        return {
            'product_tmpl_id': self.product_id.product_tmpl_id.id,
            'name': self.wizard_id.invoice_id.supplier_partner_id.id,
            'min_qty': 0.0,
            'delay': 1,
        }
```

What about use result from `super` and change to custom values required in pro of easiest inheriting?

`res = supers(...); res.update({...}); res.pop()`

(I don't know maybe is not possible)

**Review by moylop260 — APPROVED** [review]
👍 
Thanks!

Just small comments non-blocking

File: account_invoice_supplierinfo_update_discount/models/account_invoice_line.py
Comment (legalsylvain) [reply account_invoice_supplierinfo_update_discount/models/account_invoice_line.py]: Your proposal doesn't work.

bool_error

But the following line does work.

        res['current_discount'] = partnerinfo and partnerinfo.discount

**Comment by legalsylvain** [comment]
@moylop260 : @chafique-delli and me updated our PR, according your last remarks.
Is it now all ok for you ? 

kind regards.

**Comment by legalsylvain** [comment]
Ping @moylop260 !

(And happy New year !)

**Comment by legalsylvain** [comment]
Previous review was OK. 3 :+1:. merging.

**Pull request merged**
